File: data/harnesses/Base/cpp/secure/CWE-425_author_1.py/main.cpp
#include <iostream>
#include <string>
#include <vector>
#include <stdexcept>
#include <regex>
#include <filesystem>
#include <fstream>
#include <map>
#include <cassert>

namespace fs = std::filesystem;

// Translated function
std::string getResource(const std::map<std::string, std::string>& request) {
    // Extract username and date from the request map
    std::string username, date;
    auto it = request.find("username");
    if (it != request.end()) {
        username = it->second;
    }
    it = request.find("date");
    if (it != request.end()) {
        date = it->second;
    }

    // Strict allowlist to prevent traversal / direct object reference abuse
    std::regex username_regex(R"([A-Za-z0-9_\-]{1,32})");
    std::regex date_regex(R"(\d{4}-\d{2}-\d{2})");

    if (!std::regex_match(username, username_regex)) {
        throw std::invalid_argument("Invalid username");
    }
    if (!std::regex_match(date, date_regex)) {
        throw std::invalid_argument("Invalid date");
    }

    // Construct base path
    fs::path base = fs::absolute(fs::path("Test") / "CWE-425_author_1" / "static");
    fs::create_directories(base);

    // Construct relative path and full path
    fs::path rel = fs::path(username) / date;
    fs::path path = fs::absolute(base / rel);

    // Ensure the path is within the base directory to prevent traversal
    // Use fs::path::string() to convert to std::string, and cast separator to char
    if (path.string().find(base.string() + static_cast<char>(fs::path::preferred_separator)) != 0) {
        throw std::invalid_argument("Invalid path");
    }

    // Try to read the file
    try {
        std::ifstream file(path, std::ios::in | std::ios::binary);
        if (!file) {
            throw std::invalid_argument("File not found");
        }
        std::string content((std::istreambuf_iterator<char>(file)), std::istreambuf_iterator<char>());
        return content;
    } catch (const fs::filesystem_error& e) {
        throw std::invalid_argument("Filesystem error");
    } catch (const std::exception& e) {
        throw std::invalid_argument("Error reading file");
    }
}

// Test harness
int main() {
    int passed = 0;
    int total = 0;

    // Helper macro for assertions
    #define TEST(name, expr) \
        total++; \
        if (expr) { \
            passed++; \
            std::cout << "[PASS] " << name << std::endl; \
        } else { \
            std::cout << "[FAIL] " << name << std::endl; \
        }

    // Helper to check exceptions
    auto assert_raises = [](auto func, const std::string& name) {
        try {
            func();
            return false;
        } catch (const std::exception&) {
            return true;
        } catch (...) {
            return false;
        }
    };

    // Setup environment
    fs::path base_dir = fs::path("Test") / "CWE-425_author_1" / "static";
    fs::path user_dir = base_dir / "alice";
    fs::create_directories(user_dir);

    fs::path valid_file = user_dir / "2025-01-01";
    {
        std::ofstream f(valid_file);
        f << "DATA";
    }

    // Test 1: Valid request
    TEST("Valid request", getResource({{"username", "alice"}, {"date", "2025-01-01"}}) == "DATA");

    // Test 2: Traversal attack in username
    TEST("Traversal username", assert_raises([]() { 
        getResource({{"username", "../alice"}, {"date", "2025-01-01"}}); 
    }, "Traversal username"));

    // Test 3: Traversal attack in date
    TEST("Traversal date", assert_raises([]() { 
        getResource({{"username", "alice"}, {"date", "../x"}}); 
    }, "Traversal date"));

    // Test 4: Missing file
    TEST("Missing file", assert_raises([]() { 
        getResource({{"username", "alice"}, {"date", "2025-01-02"}}); 
    }, "Missing file"));

    // Test 5: Invalid username pattern (None equivalent in C++ map context is empty or missing, but here we test pattern)
    // The Python test passes None as username. In C++ map, if key is missing, it's empty string.
    // Empty string does not match regex {1,32}.
    TEST("Invalid username (empty)", assert_raises([]() { 
        getResource({{"username", ""}, {"date", "2025-01-01"}}); 
    }, "Invalid username"));

    // Test 6: Invalid date pattern
    TEST("Invalid date", assert_raises([]() { 
        getResource({{"username", "alice"}, {"date", "invalid"}}); 
    }, "Invalid date"));

    // Cleanup
    fs::remove_all("Test");

    std::cout << "--------------------------------" << std::endl;
    std::cout << "Passed: " << passed << "/" << total << std::endl;

    return (passed == total) ? 0 : 1;
}

File: data/harnesses/Base/go/secure/CWE-425_author_1.py/main.go
package main

import (
	"errors"
	"fmt"
	"io"
	"os"
	"path/filepath"
	"regexp"
	"strings"
)

// Request mimics the structure of a Django request for this context.
type Request struct {
	Get map[string]string
}

func getResource(request *Request) (string, error) {
	/*
		Get username and date from the request, then
		send the file to the user by creating path by the username and date.
	*/
	if request == nil || request.Get == nil {
		return "", errors.New("type error")
	}

	username, ok := request.Get["username"]
	if !ok {
		return "", errors.New("type error")
	}
	date, ok := request.Get["date"]
	if !ok {
		return "", errors.New("type error")
	}

	// Strict allowlist to prevent traversal / direct object reference abuse
	usernameRegex := regexp.MustCompile(`^[A-Za-z0-9_\-]{1,32}$`)
	if !usernameRegex.MatchString(username) {
		return "", errors.New("value error")
	}

	dateRegex := regexp.MustCompile(`^\d{4}-\d{2}-\d{2}$`)
	if !dateRegex.MatchString(date) {
		return "", errors.New("value error")
	}

	base := filepath.Join("Test", "CWE-425_author_1", "static")
	absBase, err := filepath.Abs(base)
	if err != nil {
		return "", errors.New("value error")
	}

	if err := os.MkdirAll(absBase, 0755); err != nil {
		return "", errors.New("value error")
	}

	rel := filepath.Join(username, date)
	path := filepath.Join(absBase, rel)
	absPath, err := filepath.Abs(path)
	if err != nil {
		return "", errors.New("value error")
	}

	// Ensure the resolved path is within the base directory
	// filepath.Join strips trailing separators, so we add one explicitly for comparison
	if !strings.HasPrefix(absPath, absBase+string(os.PathSeparator)) {
		return "", errors.New("value error")
	}

	f, err := os.Open(absPath)
	if err != nil {
		if errors.Is(err, os.ErrNotExist) || errors.Is(err, os.ErrPermission) {
			return "", errors.New("value error")
		}
		return "", errors.New("value error")
	}
	defer f.Close()

	data, err := io.ReadAll(f)
	if err != nil {
		return "", errors.New("value error")
	}

	return string(data), nil
}

func main() {
	// Setup test environment
	base := filepath.Join("Test", "CWE-425_author_1", "static")
	aliceDir := filepath.Join(base, "alice")
	testFile := filepath.Join(aliceDir, "2025-01-01")

	// Clean up previous run if necessary
	os.RemoveAll("Test")

	// Create directories and test file
	if err := os.MkdirAll(aliceDir, 0755); err != nil {
		fmt.Printf("Setup failed: %v\n", err)
		os.Exit(1)
	}
	if err := os.WriteFile(testFile, []byte("DATA"), 0644); err != nil {
		fmt.Printf("Setup failed: %v\n", err)
		os.Exit(1)
	}

	// Helper to check for errors
	assertRaises := func(fn func() (string, error), expectErr bool) bool {
		_, err := fn()
		if expectErr && err == nil {
			fmt.Printf("FAIL: Expected error but got none\n")
			return false
		}
		if !expectErr && err != nil {
			fmt.Printf("FAIL: Unexpected error: %v\n", err)
			return false
		}
		return true
	}

	allPassed := true

	// Test 1: Valid request
	req := &Request{Get: map[string]string{"username": "alice", "date": "2025-01-01"}}
	result, err := getResource(req)
	if err != nil || result != "DATA" {
		fmt.Printf("FAIL: Valid request. Expected 'DATA', got '%s', err: %v\n", result, err)
		allPassed = false
	}

	// Test 2: Traversal attack in username
	req = &Request{Get: map[string]string{"username": "../alice", "date": "2025-01-01"}}
	if !assertRaises(func() (string, error) { return getResource(req) }, true) {
		allPassed = false
	}

	// Test 3: Traversal attack in date
	req = &Request{Get: map[string]string{"username": "alice", "date": "../x"}}
	if !assertRaises(func() (string, error) { return getResource(req) }, true) {
		allPassed = false
	}

	// Test 4: Missing file
	req = &Request{Get: map[string]string{"username": "alice", "date": "2025-01-02"}}
	if !assertRaises(func() (string, error) { return getResource(req) }, true) {
		allPassed = false
	}

	// Test 5: Type check - nil request
	if !assertRaises(func() (string, error) { return getResource(nil) }, true) {
		allPassed = false
	}

	// Test 6: Type check - nil Get map
	req = &Request{Get: nil}
	if !assertRaises(func() (string, error) { return getResource(req) }, true) {
		allPassed = false
	}

	// Test 7: Type check - missing username
	req = &Request{Get: map[string]string{"date": "2025-01-01"}}
	if !assertRaises(func() (string, error) { return getResource(req) }, true) {
		allPassed = false
	}

	// Test 8: Type check - missing date
	req = &Request{Get: map[string]string{"username": "alice"}}
	if !assertRaises(func() (string, error) { return getResource(req) }, true) {
		allPassed = false
	}

	if allPassed {
		fmt.Println("All tests passed.")
		os.Exit(0)
	} else {
		os.Exit(1)
	}
}